File: apps/server/src/udp/forzaUdpReceiver.ts
import dgram from "node:dgram";
import { ForzaPacketParser } from "../parser/forzaPacketParser.js";
import { TelemetryStore } from "../telemetry/telemetryStore.js";

export type ForzaUdpReceiverOptions = {
  host: string;
  port: number;
  parser: ForzaPacketParser;
  store: TelemetryStore;
};

export function startForzaUdpReceiver(options: ForzaUdpReceiverOptions): dgram.Socket {
  const socket = dgram.createSocket("udp4");

  // UDP receive is intentionally immediate: every packet is parsed as it arrives
  // and only the latest in-memory telemetry snapshot is replaced.
  socket.on("message", (packet, remote) => {
    try {
      const snapshot = options.parser.parse(packet);
      options.store.update(snapshot, options.parser.getLastDiagnostics());
    } catch (error) {
      options.store.setLastPacketInfo({
        length: packet.length,
        format: "parse-error",
        profile: "parse-error",
        dashShift: null,
        accepted: false,
        errors: [error instanceof Error ? error.message : String(error)]
      });
      console.error("[udp] Failed to parse Forza telemetry packet", {
        remote: `${remote.address}:${remote.port}`,
        length: packet.length,
        error
      });
    }
  });

  socket.on("error", (error) => {
    console.error("[udp] Socket error", error);
  });

  socket.bind(options.port, options.host, () => {
    const address = socket.address();
    console.log(`[udp] Listening for Forza Data Out on ${address.address}:${address.port}`);
  });

  return socket;
}

File: apps/server-rs/src/udp/forza_udp_receiver.rs
use tokio::{
    net::UdpSocket,
    select,
    sync::watch,
    time::{Duration, Instant},
};

use crate::{
    parser::forza_packet_parser::ForzaPacketParser,
    telemetry::{
        telemetry_broadcaster::TelemetryBroadcaster, telemetry_store::TelemetryStore,
        telemetry_types::TelemetryPacketInfo,
    },
};

const PACKET_INFO_SAMPLE_INTERVAL: Duration = Duration::from_millis(250);

pub async fn run_forza_udp_receiver(
    socket: UdpSocket,
    store: TelemetryStore,
    parser: ForzaPacketParser,
    broadcaster: TelemetryBroadcaster,
    mut stop_rx: watch::Receiver<bool>,
) -> std::io::Result<()> {
    tracing::info!(
        address = %socket.local_addr()?,
        "listening for Forza Data Out UDP packets"
    );

    let mut packet = vec![0_u8; 2048];
    let mut next_packet_info_sample_at = Instant::now();

    loop {
        // UDP receive stays on the hot path: parse immediately, overwrite the
        // latest in-memory snapshot, then request a capped WebSocket broadcast.
        let receive_result = select! {
            changed = stop_rx.changed() => {
                if changed.is_err() || *stop_rx.borrow() {
                    tracing::info!("UDP receiver stop requested");
                    break;
                }
                continue;
            }
            result = socket.recv_from(&mut packet) => result,
        };

        let (length, remote) = receive_result?;
        let packet_slice = &packet[..length];
        let capture_packet_info = Instant::now() >= next_packet_info_sample_at;

        let parse_result = if capture_packet_info {
            parser
                .parse(packet_slice)
                .map(|(snapshot, packet_info)| (snapshot, Some(packet_info)))
        } else {
            parser
                .parse_snapshot(packet_slice)
                .map(|snapshot| (snapshot, None))
        };

        match parse_result {
            Ok((snapshot, packet_info)) => {
                if capture_packet_info {
                    next_packet_info_sample_at = Instant::now() + PACKET_INFO_SAMPLE_INTERVAL;
                }
                let snapshot = store.update(snapshot, packet_info).await;
                broadcaster.request_broadcast(snapshot);
            }
            Err(error) => {
                store
                    .set_last_packet_info(TelemetryPacketInfo {
                        length,
                        format: Some("parse-error".to_string()),
                        profile: "parse-error".to_string(),
                        dash_shift: None,
                        accepted: false,
                        errors: vec![error.clone()],
                        candidates: None,
                    })
                    .await;

                tracing::error!(
                    remote = %remote,
                    length,
                    error,
                    "failed to parse Forza telemetry packet"
                );
            }
        }
    }

    Ok(())
}
